File: Sean的leetcode/Leetcode/Word Squares.cpp
class Solution {
public:
    vector<vector<string>> wordSquares(vector<string>& words) {
        int n = words.size();
        vector<vector<string>> squares;
        if (n == 0) return squares;
        int size = words[0].size();
        vector<string> square(size);
        unordered_map<string, vector<string>> table;
        for (auto word : words) {
            for (int i = 0; i < word.size(); i++) {
                table[word.substr(0, i)].push_back(word);
            }
        }
        build(table, square, squares, 0, size);
        return squares;
    }
    void build(unordered_map<string, vector<string>>& table, vector<string>& square, vector<vector<string>>& squares, int i, int target) {
        if (i == target) {
            squares.push_back(square);
            return;
        }
        string prefix;
        for (int k = 0; k < i; k++) {
            prefix += square[k][i];
        }
        for (auto word: table[prefix]) {
            square[i] = word;
            build(table, square, squares, i+1, target);
        }
    }
};

File: MyLeetCode/python/Word Squares.py
###
# Back Tracking
# Time Complexity: O(n*l) build a trie/hash table  + O(n^l) (use pruning, hard to analyze)
###
class TrieNode(object):
    def __init__(self):
        self.children = {}
        self.startwith = []
        
class Trie(object):
    def __init__(self, words):
        self.root = TrieNode()
        for word in words:
            cur = self.root
            for c in word:
                if c not in cur.children:
                    cur.children[c] = TrieNode()
                cur.startwith.append(word)
                cur = cur.children[c]
                
    def find_by_prefix(self, prefix):
        cur = self.root
        for c in prefix:
            if c not in cur.children:
                return []
            cur = cur.children[c]
        return cur.startwith
    
class Solution(object):
    def wordSquares(self, words):
        """
        :type words: List[str]
        :rtype: List[List[str]]
        """
        if not words:
            return []
        size = len(words[0])
        trie = Trie(words)
        square = []
        squares = []
        self.bt(trie, 0, size, square, squares)
        return squares
    
    def bt(self, trie, cur, target, square, squares):
        if len(square) == target:
            squares.append(square)
            return
        prefix = ''
        if cur:
            prefix = ''.join(zip(*square)[cur])
        for word in trie.find_by_prefix(prefix):          
            self.bt(trie, cur + 1, target, square + [word], squares)
            

class Solution(object):
    
    def wordSquares(self, words):
        """
        :type words: List[str]
        :rtype: List[List[str]]
        """
        from collections import defaultdict
        if not words:
            return []
        size = len(words[0])
        table = defaultdict(list)
        for word in words:
            for i in xrange(size):
                table[word[0:i]].append(word)
        squares = []
        square = [] 
        self.build(table, size, square, squares)
        return squares
    def build(self, table, size, square, squares):
        if len(square) == size:
            squares.append(square)
            return
        prefix = ''
        if square:
            prefix = ''.join(zip(*square)[len(square)])
        for word in table[prefix]:
            self.build(table, size, square + [word], squares)
        
        